File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
         int leftWeight = 0, rightWeight = 0;
        for (int weight : weights) {
            leftWeight = Math.max(leftWeight, weight);
            rightWeight += weight;
        }
          
        while (leftWeight < rightWeight) {
            int midWeight = leftWeight + (rightWeight - leftWeight) / 2;
            
            int need = 1, cur = 0;
            for (int weight : weights) {
                if (cur + weight > midWeight) {
                    need++;
                    cur = 0;
                }
                cur += weight;
            }
            
            if (need > days)leftWeight = midWeight + 1;
             else  rightWeight = midWeight;
            
        }
        return leftWeight;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        
        def canGetAllDays(w,d,randN):
            t=0
            ud=1
            for ew in w:
                t+=ew
                if t>randN:
                    t=ew
                    ud+=1
            return ud<=d
        L=max(weights)
        R=sum(weights)
        while L<=R:
            M=(L+R)//2
            if(canGetAllDays(weights,days, M)):
                R=M-1
            else:
                L=M+1
        return L
        